x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/process/PyTorchStateStreamer.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.pytorch.process;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.client.OriginSettingClient;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.xpack.ml.inference.persistence.ChunkedTrainedModelRestorer;
+import org.elasticsearch.xpack.ml.inference.persistence.TrainedModelDefinitionDoc;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.concurrent.ExecutorService;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+
+/**
+ * PyTorch models in the TorchScript format are binary files divided
+ * into small chunks and base64 encoded for storage in Elasticsearch.
+ * The model is restored by base64 decoding the stored state and streaming
+ * the binary objects concatenated in order. There is no delineation between
+ * individual chunks the state should appear as one contiguous file.
+ */
+public class PyTorchStateStreamer {
+
+    private static final Logger logger = LogManager.getLogger(PyTorchStateStreamer.class);
+
+    private final OriginSettingClient client;
+    private final ExecutorService executorService;
+    private final NamedXContentRegistry xContentRegistry;
+    private volatile boolean isCancelled;
+    private boolean modelSizeWritten = false;
+
+    public PyTorchStateStreamer(Client client, ExecutorService executorService, NamedXContentRegistry xContentRegistry) {
+        this.client = new OriginSettingClient(Objects.requireNonNull(client), ML_ORIGIN);
+        this.executorService = Objects.requireNonNull(executorService);
+        this.xContentRegistry = Objects.requireNonNull(xContentRegistry);
+    }
+
+    /**
+     * Cancels the state streaming at the first opportunity.
+     */
+    public void cancel() {
+        isCancelled = true;
+    }
+
+    /**
+     * First writes the size of the model so the native process can
+     * allocated memory then writes the chunks of binary state.
+     *
+     * @param modelId  The model to write
+     * @param restoreStream The stream to write to
+     * @param listener  error and success listener
+     */
+    public void writeStateToStream(String modelId, OutputStream restoreStream, ActionListener<Boolean> listener) {
+        ChunkedTrainedModelRestorer restorer = new ChunkedTrainedModelRestorer(modelId, client, executorService, xContentRegistry);
```

Comment:
I wonder if we should set the search size to 1 here.

At the moment with the default search size of 10:

- We load 160MB of UTF-8 base64 encoded strings (10 * 16MB).
- Then we parse one of these into a UTF-16 base64 encoded string, so that's an extra 32MB.
- While those two formats are held in memory, we then convert back to UTF-8 in a way that doesn't reuse the original UTF-8 representation, so that's an extra 16MB.
- Then we create the raw bytes, which on average would be about 1/4 of the size, so 4MB.

So all that adds to 212MB.

ML nodes in Cloud have JVM heaps capped at 2GB, so that's a big chunk to be using for a single operation.

For the anomaly detector model state we take care not to hold the same data in multiple formats simultaneously:

- We load one doc at a time
- We never convert the UTF-8 base64 encoded data to UTF-16
- We stream it to the C++ which has a streaming base64 decoder so also never holds the full data in two formats

So in this case we only use the 16MB of a single model state document in UTF-8 format.

A sliding scale of simplicity and return on investment for reducing the memory usage of restoration is:

1. Search for 1 document at a time
2. Change `TrainedModelDefinition` so that it stores the model definition in a UTF-8 `byte[]` rather than UTF-16 `String`
3. Change `TrainedModelDefinition` so that it stores the model definition in a raw `byte[]` rather than UTF-16 `String` (decoded directly from the UTF-8 of `_source` rather than going via an intermediate temporary `String`)

2 and 3 are non-trivial pieces of work but 1 is a one-liner so could be added to this PR.

I realise we already have this problem with the `TrainedModelProvider` class.  I expect we've never hit it in practice because our boosted tree models are in reality considerably smaller than the maximum.  But these PyTorch models seem to be bigger more often, so it would be good to consider memory efficiency from the start.

Reply:
++ 

2 might actually be a viable option. `XContentGenerator` has a `writeBinary(byte[] value)` method which calls a Jackson `writeBinary(byte[] value)` which in turn base64 encodes the bytes. `XContentParser` does the reverse with `byte[] binaryValue()`. For BWC the binary data could be written to a new field and when parsing the internal content would be created from either the old string field or the new binary one. 

There is a complication that when the model is PUT it is already base64 encoded and would have to be un-encoded before storage otherwise it would be double base64 encoded. The savings would be significant though and worth investigating for a follow up PR

Reply:
I think this is a good idea as well!

Wish I thought of it 🤦